File: Client/src/data/json/JsonParser.cpp
//
// Created by Nathan David on 26/01/2022.
//

#include "data/json/JsonParser.h"

pair<DataObject *, int> JsonParser::parseObject(string input) {
    auto *object = new DataObject();
    Token token;
    unsigned long length = input.length();
    do {
        // Pour la clé du token on récupère le token actuel
        token = JsonToken::getNextToken(input);
        input = input.substr(token.position); // On avance à la fin de ce token
        string key = token.value;

        if (token.type == TOKEN::CURLY_CLOSE || token.type == TOKEN::UNKNOWN) {
            break;
        }

        // Pour récupérer la valeur, on va voir le prochain token
        token = JsonToken::getNextToken(input);

        switch (token.type) {
            case TOKEN::STRING: {
                object->put(key, new DataPrimitive(token.value));
                input = input.substr(token.position); // On avance à la fin de la valeur
                break;
            }
            case TOKEN::CURLY_OPEN: {
                input = input.substr(
                        JsonToken::getNextTokenPosition(input, '{') + 1); // On avance au début de ce token
                pair<DataObject *, int> res = parseObject(input);
                object->put(key, res.first);
                input = input.substr(res.second + 1); // On avance à la fin de la liste

                break;
            }
            case TOKEN::ARRAY_OPEN: {
                input = input.substr(
                        JsonToken::getNextTokenPosition(input, '[') + 1); // On avance au début de ce token
                pair<DataArray *, int> res = parseArray(input);
                object->put(key, res.first);
                input = input.substr(res.second + 1); // On avance à la fin de la liste

                break;
            }
            default:
                break;
        }


    } while (token.type != TOKEN::CURLY_CLOSE && token.type != TOKEN::UNKNOWN);

    return {object, length - input.length()};
}

string JsonParser::toString(const DataObject *object) const {
    string res = "{";

    for (const auto &it: object->getChildren()) {
        DataElement *element = it.second;
        res += '"' + it.first + "\":";
        res += element->toString();

        res += ",";
    }

    if (res.length() > 1)
        res.pop_back();

    res += "}";
    return res;
}

string JsonParser::toString(const DataArray *array) const {
    string res = "[";

    for (auto it: *array) {
        DataElement *element = it;
        res += element->toString();

        res += ",";
    }


    if (res.length() > 1)
        res.pop_back();

    res += "]";
    return res;
}

string JsonParser::toString(const DataPrimitive *primitive) const {
    return '"' + primitive->getAsString() + '"';
}

File: Server/src/main/java/fr/nathan/mim/api/data/json/JsonParser.java
package fr.nathan.mim.api.data.json;

import fr.nathan.mim.api.Pair;
import fr.nathan.mim.api.data.*;

import java.util.Map;

public class JsonParser extends Parser {

    @Override
    public DataObject parse(String input) {
        return parseObject(input.substring(1)).getFirst();
    }

    private Pair<DataArray, Integer> parseArray(String input) {
        DataArray array = new DataArray();
        JsonTokenParser token;
        long length = input.length();
        do {
            token = JsonTokenParser.getNextToken(input);
            input = input.substring(token.getPosition()); // On avance à la fin de ce token

            if (token.getType() == JsonToken.ARRAY_CLOSE || token.getType() == JsonToken.UNKNOWN) {
                break;
            }

            switch (token.getType()) {
                case STRING: {
                    array.add(new DataPrimitive(token.getValue()));
                    input = input.substring(token.getPosition()); // On avance à la fin de la valeur
                    break;
                }
                case CURLY_OPEN: {
                    input = input.substring(
                            JsonTokenParser.getNextTokenPosition(input, '{') + 1); // On avance au début de ce token
                    Pair<DataObject, Integer> res = parseObject(input);
                    array.add(res.getFirst());
                    input = input.substring(res.getSecond() + 1); // On avance à la fin de l'objet

                    break;
                }
                case ARRAY_OPEN: {
                    input = input.substring(JsonTokenParser.getNextTokenPosition(input, '[') + 1); // On avance au début de ce token
                    Pair<DataArray, Integer> res = parseArray(input);
                    array.add(res.getFirst());
                    input = input.substring(res.getSecond() + 1); // On avance à la fin de la liste

                    break;
                }
            }

        } while (token.getType() != JsonToken.ARRAY_CLOSE && token.getType() != JsonToken.UNKNOWN);

        return new Pair<>(array, (int) length - input.length());
    }

    private Pair<DataObject, Integer> parseObject(String input) {

        DataObject object = new DataObject();
        JsonTokenParser token;
        long length = input.length();
        do {
            // Pour la clé du token on récupère le token actuel
            token = JsonTokenParser.getNextToken(input);
            input = input.substring(token.getPosition()); // On avance à la fin de ce token
            String key = token.getValue();

            if (token.getType() == JsonToken.CURLY_CLOSE || token.getType() == JsonToken.UNKNOWN) {
                break;
            }

            // Pour récupérer la valeur, on va voir le prochain token
            token = JsonTokenParser.getNextToken(input);
            switch (token.getType()) {
                case STRING: {
                    object.put(key, new DataPrimitive(token.getValue()));
                    input = input.substring(token.getPosition()); // On avance à la fin de la valeur
                    break;
                }
                case CURLY_OPEN: {
                    input = input.substring(
                            JsonTokenParser.getNextTokenPosition(input, '{') + 1); // On avance au début de ce token
                    Pair<DataObject, Integer> res = parseObject(input);
                    object.put(key, res.getFirst());
                    input = input.substring(res.getSecond() + 1); // On avance à la fin de la liste

                    break;
                }
                case ARRAY_OPEN: {
                    input = input.substring(
                            JsonTokenParser.getNextTokenPosition(input, '[') + 1); // On avance au début de ce token
                    Pair<DataArray, Integer> res = parseArray(input);
                    object.put(key, res.getFirst());
                    input = input.substring(res.getSecond() + 1); // On avance à la fin de la liste

                    break;
                }
            }

        } while (token.getType() != JsonToken.CURLY_CLOSE && token.getType() != JsonToken.UNKNOWN);

        return new Pair<>(object, (int) length - input.length());
    }

    //////

    @Override
    public String toString(DataObject object) {
        StringBuilder res = new StringBuilder("{");

        for (Map.Entry<String, DataElement> entry : object.getChildren().entrySet()) {
            res.append('"').append(entry.getKey()).append("\":");
            res.append(entry.getValue().toString());
            res.append(",");
        }

        if (res.length() > 1)
            res.setLength(res.length() - 1);

        res.append("}");
        return res.toString();
    }

    @Override
    public String toString(DataArray array) {
        StringBuilder res = new StringBuilder("[");

        for (DataElement element : array.getElements()) {
            res.append(element.toString());
            res.append(",");
        }


        if (res.length() > 1)
            res.setLength(res.length() - 1);

        res.append("]");
        return res.toString();
    }

    @Override
    public String toString(DataPrimitive primitive) {
        return '"' + primitive.getAsString() + '"';
    }
}
